README
MCrawler version 0.01
=====================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL				 --> creates make file
   make                          --> automatically checks dependencies and install required packages from CPAN
   make test					 --> tests not working need to write some
   make install					 --> avoid installing it since in development phase
   
   Creating Database
   
   createdb -W -O crawler -U crawler crawlerdb   			   --> crawler is user
   												 			   --> crawlerdb is DB name
   												 			   --> default password is 'crawler'
   
   All default database values can be changed in MCrawler_config.yml
   
   psql -f <root_dir>/MCrawler_DB.sql -U crawler -d crawlerdb  --> creates database
   
   create roles for database clients:
   login psql with super user permissions
   CREATE ROLE "client_"<client_id> WITH LOGIN PASSWORD <client_database_password> ;
   example: CREATE ROLE client_8989 WITH LOGIN PASSWORD 'secret';
   
   Starting the Crawler and Server 
   
   cd <Root_directory>
   perl Crawler.PL
   
   Starting Client
   
   perl MClient.PL
   
   Client command format:
   Making new request   		 --> new_request
   Checking messages from server --> check_inbox
   Queueing a URL				 --> queue_url <request_id> <url>
   Queueing a file with URLs	 --> queue_file <request_id> <file_location>              
   Other Settings				 --> downloads_type <request_id> <user_value>			  not_used
   								 --> depth_of_search <request_id> <user_value>			  default_value -> 3
   								 --> refresh_rate <request_id> <user_value>				  default_value -> 60*60*24
   								 --> allowed_content <request_id> <user_value>			  default_value -> html|txt
   								 --> user_agent <request_id> <user_value>				  default_value -> MCrawler bot (http://cyber.law.harvard.edu)
								 --> dequeue_request <request_id>
								 --> check_status <request_id>
   
   format of seeds.txt file is new line seperated URLs 
DEPENDENCIES

This module requires these other modules and libraries:

	These are compulsory modules inorder to run Makefile.PL
	
	-> ExtUtils::MakeMaker
	-> ExtUtils::AutoInstall
	
SAMPLE INPUT/OUTPUT
Admin creates a role by logging into crawler database with psql
CREATE ROLE client_8989 WITH LOGIN PASSWORD 'client_password';
change server port(default is 6666) and client id in MClient.PL pass it to client
client starts a request by entering new_request.here is a series of standard commands.

--> initiating a new request
input --> new_request
input --> check_inbox
output --> you(8989) made new request 23

--> sending input to server
input --> queue_url 23 http://wikipedia.org
input --> queue_file 23 ./seeds.txt

--> configuring the request
input --> depth_of_search 23 10
input --> refresh_rate 23 7200
input --> user_agent 23 MCrawler-bot/1.0 (http://cyber.law.harvard.edu)

--> committing the request 
input --> commit_request 23

-->status checking
input --> check_status 23
output --> please check your request status on view_8989_23 table.
			Out of 1 URLs,0 URLs are completed and remaining are in processing.

next client login to postgres server with given password by admin.
and can only has SELECT permission on view_8989_23 view.

-->dequeueing
input --> dequeue_request 23 
dequeue removes request_id from the queue and stops any requests with that request_id.
And any data in database with that request_id will be removed.


if client has module to contact server directly syntax of communication can be seen after executing every command.

COPYRIGHT AND LICENCE

Copyright (C) 2010 by aditya

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension MCrawler.

0.01  Fri Jul 16 01:09:39 2010
	- original version; created by h2xs 1.23 with options
		-AXc -n MCrawler



MClient.pl
#!/usr/bin/perl
use constant SERVER_PORT => 6666;
use constant CLIENT_ID => 8989;
use strict;
use warnings;
use Switch;
use IO::Socket::INET;
use threads;
use threads::shared;
use Perl6::Say;

# Create a new socket

my $MySocket;
my $def_msg="Enter message to send to server : ";
my $server_reply="";
my $inbox : shared ="";

sub connect_server{
my $MySocket=new IO::Socket::INET->new(	PeerPort=>SERVER_PORT,
										Proto=>'tcp',
										PeerAddr=>'localhost'
									   );
die "Could not create socket: $!\n" unless $MySocket;
return $MySocket;
}

sub send_to_server{
	my $msg = CLIENT_ID." ".$_[0]."\n";
	#print STDERR $msg;
		if(send($MySocket,$msg,0))
		{
			return "sent $_[0] ";
		}
		else
		{
			return "error in sending message";
		}
}

sub handle_queue{
	if(@_ != 0){
	my $request_id = shift(@_);
	my $input = join(" ",@_);
	return send_to_server($request_id." queue ".$input);
	}
}

sub load_file_and_queue{
	my $request_id = shift(@_);
	my $file = shift(@_);
	my $data = " ";
	if(open LOAD_FILE, "<$file")
	{
		while (<LOAD_FILE>) 
		{
			my $url= $_;
			chomp($url);
			$data = $data." ".$url;
		}
#		print ("queue".$data);
		return send_to_server($request_id." queue ".$data);
	}
	else 
	{
		return "Could not open file '$file '. $!" ;
	}
}

sub listen_server{
		while ($server_reply = <$MySocket>)
			{
			if($server_reply ne '')
				{
				$inbox = $inbox."\n".$server_reply;
				}
			}	
}

sub start_client{
	print "client started ";
	$MySocket = connect_server();
	print "connected to server\n";
	my $thr = threads->new(\&listen_server);
	print "enter 'help' for help \n";
		while (1)
			{	print "\nSend message to server : ";
				my $input = <STDIN>;
				$input =~s/^\s+|\s+$|\n//g;
				my @tokens = split(/\s+/,$input);
				#print STDERR $tokens[0];
				#my @case_hash = {"downloads_type","depth_of_search","refresh_rate","allowed_content","user_agent","commit_request"};
				switch ($tokens[0])
				{
					my $token = shift(@tokens);
					case "help"{
						            print " \n The follwing are the commands:\n new_request --> initializing a new request \n check_inbox --> check messages from server for any new request_id\n queue_url <request_id><space><url> --> queueing a single url\n queue_file <request_id><space><file_location> --> queueing a file with urls\n allowed_content <request_id><space><value> --> setting content type default - html|text\n depth_of_search <request_id><space><value> --> setting depth of search default - 3 \n refresh_rate <request_id><space><value> --> setting refresh rate default - 24*24*3600\n user_agent <request_id><space><value> --> setting user_agent default -\n commit_request <request_id> --> starts the defined request\n dequeue_request -->dequeue request removes all data regarding that request from database\n check_status -->gives present status of the request ";
					}					
					case "new_request"{
									print send_to_server("request");
					}
					case "queue_url" {
									print handle_queue(@tokens);  
					}
					case "queue_file"{
									print load_file_and_queue(@tokens);
					}
					case["downloads_type","depth_of_search","refresh_rate","allowed_content","user_agent","check_status","commit_request","dequeue_request"]
					{
						my $request_id=shift(@tokens);
						if (defined($request_id)) {
									print send_to_server($request_id." ".$token." ".join(" ",@tokens) );	
									}
					    else{
					    			print "please enter with request_id\ncheck syntax with help command";
					    			}
									
					}
					case "check_inbox"{
									print "\n messages received from server are: ".$inbox;
									$inbox = "";
					}
					else{
						print "wrong input \ntype 'help' for help\n";
					}
				}

			}
				
}

start_client();
